### PR TITLE
[PB-1482]: feat/compatibility-with-mongo-db-to-4.4

### PR DESCRIPTION
**Pull request by sw-wayner**
# Add end-to-end tests (e2e)
In order to upgrade `mongo` from version `3.6` to `4.4`, e2e tests where added to ensure the main features work fine

## Used strategy 

1.  Add supertest package
2. Add configuration for e2e testing
3.  Add e2e tests to cover:
    - [x] Upload/download/share/delete files
    - [x] Create/Update/Delete buckets
    - [x] Create/Update/Delete users

> [!NOTE] 
> Related task: [PB-1482](https://inxt.atlassian.net/browse/PB-1482)

[PB-1482]: https://inxt.atlassian.net/browse/PB-1482?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by sg-gs** [comment]
Do not count with #120 @sw-wayner. The code could be useful for this PR, but it is not updated to the actual codebase, so assume everything would be packed here. We will close that PR for being stalled soon.

File: tests/lib/e2e/utils.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+}
+
+export const getAuth = (user: { email: string , hashpass: string }) => {
+  const credential = Buffer.from(`${user.email}:${user.hashpass}`).toString('base64');
+  return `Basic ${credential}`;
+}
+
+
+export const cleanDataBase = (storage: any) => {
```
Comment (sg-gs) [inline tests/lib/e2e/utils.ts]: Make this `async`, you should `await` for it, if not, the promises would be left there, that is a leak.

File: infrastructure/mongo-init.js [hunk 2]
```diff
@@ -3,7 +3,8 @@ db.createUser({
   pwd: 'password',
   roles: [
     { role: 'readWrite', db: '__inxt-network' },
-    { role: 'readAnyDatabase', db: 'admin' }
+    { role: 'readAnyDatabase', db: 'admin' },
```
Comment (sg-gs) [inline infrastructure/mongo-init.js:6]: Why do you need this?

File: tests/lib/e2e/e2e-spec.ts [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+  beforeAll(() => {
+    cleanDataBase(engine.storage)
+  })
+
+  afterAll(() => {
+    cleanDataBase(engine.storage)
+  })
+
+  describe('Buckets', () => {
```
Comment (sg-gs) [inline tests/lib/e2e/e2e-spec.ts]: Please, adhere to the company testing guidelines that we have in the internal [docs](https://inxt.atlassian.net/wiki/spaces/T/pages/25460784/How+to+Testing+Guideline), including but not limited to the AAA pattern, which makes debugging and understanding of the test easier

**Review by sg-gs — CHANGES_REQUESTED** [review]

File: infrastructure/mongo-init.js (hunk 2 above)
Comment (sw-wayner) [reply infrastructure/mongo-init.js:6]: Without that role the `database-viewer` container fails with
```
MongoServerError: not authorized on config to execute command { listCollections: 1, filter: {}, cursor: {}, nameOnly: false, authorizedCollections: false, lsid: { id: UUID("61969390-3549-4bd0-835d-61075db37558") }, $db: "config" }
```

I added the `clusterMonitor` role to give readonly access to the database
https://www.mongodb.com/docs/v4.4/reference/built-in-roles/#mongodb-authrole-clusterMonitor

File: tests/lib/e2e/e2e-spec.ts (hunk 3 above)
Comment (sw-wayner) [reply tests/lib/e2e/e2e-spec.ts]: I added extra descriptions to the tests and organized them to follow the AAA pattern
image

File: lib/server/http/users/controller.ts [hunk 4]
```diff
@@ -110,15 +110,15 @@ export class HTTPUsersController {
...
+    const userEmail = (req as AuthorizedRequest<any>).user.email;
 
     if (!deactivator || !redirect) {
       return res.status(400).send({ error: 'Missing required params' });
     }
 
     try {
       const userRequestedToBeDestroyed = await this.usersUsecase.requestUserDestroy(
-        userId, 
+        userEmail,
```
Comment (sg-gs) [inline lib/server/http/users/controller.ts]: We should use the `user.uuid` to perform the destroy (as we will allow changing the email soon). Is there some reason to use the email instead?

File: tests/lib/e2e/users.fixtures.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+const usersTest: MongoUserModel[] = [
+  {
+    _id: v4(),
+    hashpass:
+      '4b796e7bbe57f7092d3627d3fa7e6f645b0e50e3411e1fafd61dbf27241d836d',
+    subscriptionPlan: {
+      isSubscribed: false,
+    },
+    referralPartner: null,
+    email: 'fff@ff.com',
```
Comment (sg-gs) [inline tests/lib/e2e/users.fixtures.ts]: use [chance](https://github.com/internxt/drive-server-wip/blob/59e13c6ef5a12e5aa89317e9f8eb72fd1165cfbc/test/fixtures.ts#L2) as we do use on drive-server-wip to generate random values with the most realistic format possible.

File: tests/lib/e2e/users/users.e2e-spec.ts [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+        const dbUserAfterDeactivation = await engine.storage.models.User.findOne({ _id: user.id })
+        expect(dbUserAfterDeactivation).toBeNull()
+
+      })
+
+    })
+
+    describe('User update v2', () => {
+
+      it('should be able to update a user email via gateway', async () => {
```
Comment (sg-gs) [inline tests/lib/e2e/users/users.e2e-spec.ts]: This PR still misses the product-oriented test legends, let's put an example here: 

```
describe('User updating its email', () => {
   it('When the user wants to change its email to a new free-to-use one, it works', () => {
     // some test...
   });
   
   it('When the user wants to change its email, but it already in use, it fails', () => {
    // some test
   })
   
   // and so on...
})
```

This way of writing the legends makes tests more readable and enforces it to be descriptive but not too technical if you go too much into detail, you will describe how the tested code does things, and if those things change, then you should change all the legends to be up-to-date. 

Adopting this technique also makes the test suite easy to understand and troubleshoot for other devs that may work in the future here, without knowing about the underlying implementation, but being able to spot what is failing, easily.

**Review by sg-gs — CHANGES_REQUESTED** [review]

File: lib/server/http/users/controller.ts (hunk 4 above)
Comment (sw-wayner) [reply lib/server/http/users/controller.ts]: No, I used the email because that's what `requestUserDestroy` is expecting. I'll change it.

File: tests/lib/e2e/buckets/buckets.e2e-spec.ts [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+    jest.clearAllMocks()
+  })
+
+  describe('Buckets Management', () => {
+
+    describe('Bucket Management v1', () => {
+
+      describe('Creating a bucket', () => {
+
+        it('When you want create the root bucket, it should work without any arguments', async () => {
```
Comment (sg-gs) [inline tests/lib/e2e/buckets/buckets.e2e-spec.ts]: "When you want _to_ create.."

There is a typo on multiple test legends

File: tests/lib/e2e/buckets/buckets.e2e-spec.ts [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+          // Act
+          const response = await testServer
+            .post('/buckets')
+            .set('Authorization', getAuth(testUser))
+
+          // Assert
+          expect(response.status).toBe(201)
+          expect(response.body).toHaveProperty('id')
+
+          const buckets = await engine.storage.models.Bucket.find({ _id: response.body.id })
```
Comment (sg-gs) [inline tests/lib/e2e/buckets/buckets.e2e-spec.ts]: Why not using `findOne` instead of `find`?

File: tests/lib/e2e/buckets/buckets.e2e-spec.ts [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+
+
+          // Assert
+          expect(response.status).toBe(200);
+
+          const dbBucket = await engine.storage.models.Bucket.findOne({ _id: response.body.id })
+          expect(dbBucket.toObject().pubkeys).toEqual([])
+
+        })
+        it('Whe you want to update a bucket, it should fail with invalid pubkeys list', async () => {
```
Comment (sg-gs) [inline tests/lib/e2e/buckets/buckets.e2e-spec.ts]: When, a 'n' is missing here

File: tests/lib/e2e/buckets/buckets.e2e-spec.ts [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+            .send({
+              pubkeys: ['031a259ee122414f57a63bbd6887ee17960e9106b0adcf89a298cdad2108adf4d9'],
+              name: dataGenerator.word({ length: 7 })
+            })
+            .expect(201);
+
+          // Act: Delete the bucket
+          const response = await testServer
+            .delete(`/buckets/${bucket.id}`)
+            .set('Authorization', getAuth(testUser))
```
Comment (sg-gs) [inline tests/lib/e2e/buckets/buckets.e2e-spec.ts]: Calling this user `notTheOwner` would make the test easier to read

File: tests/lib/e2e/buckets/buckets.e2e-spec.ts [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+
+
+          // Assert
+          expect(response.status).toBe(404)
+
+          await deleteTestUser({ user: owner })
```
Comment (sg-gs) [inline tests/lib/e2e/buckets/buckets.e2e-spec.ts]: Why is this required? Tests ideally should be as idempotent as possible. If you need to perform post cleanups, use `afterAll` hook

File: tests/lib/e2e/buckets/files.e2e-spec.ts [hunk 12]
```diff
@@ -0,0 +1,347 @@
...
+
+  afterEach(() => {
+    axiosGetStub.restore()
+  })
+
+  describe('File Management v2', () => {
+
+    describe('Uploading a file', () => {
+
+      it('When a user wants to upload a file, it should work for owned buckets and get a list of upload links one per each file part', async () => {
```
Comment (sg-gs) [inline tests/lib/e2e/buckets/files.e2e-spec.ts]: This test only covers multipart uploads. What about one-part uploads?

File: tests/lib/e2e/buckets/files.e2e-spec.ts [hunk 13]
```diff
@@ -0,0 +1,347 @@
...
+          .send({ uploads: [{ index: 0, size: 1000, }, { index: 1, size: 10000, },], })
+
+        // Assert
+        expect(response.status).toBe(200);
+        const { uploads } = response.body;
+
+        expect(uploads).toHaveLength(2);
+
+        const upload1 = uploads[0];
+        expect(upload1.url).toBe('http://fake-url')
```
Comment (sg-gs) [inline tests/lib/e2e/buckets/files.e2e-spec.ts]: Use a constant for storing the URL, as you have it duplicated around the code. If you want to change it, you should need to change it a lot of times, use a variable

File: tests/lib/e2e/buckets/files.e2e-spec.ts [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+          .set('Authorization', getAuth(testUser))
+          .send({ uploads: [{ index: 0, size: 1000, }, { index: 1, size: 10000, },], })
+
+        // Assert
+        expect(response.status).toBe(200);
+        const { uploads } = response.body;
+
+        expect(uploads).toHaveLength(2);
+
+        const upload1 = uploads[0];
```
Comment (sg-gs) [inline tests/lib/e2e/buckets/files.e2e-spec.ts]: Use destructuring to make code easier to read: 

```
const [firstUpload, secondUpload] = uploads;
```

File: tests/lib/e2e/buckets/files.e2e-spec.ts [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+          .send({ uploads: [{ index: 0, size: 1000, }, { index: 1, size: 10000, },], })
+
+        // Arrange: finish the upload
+        const index = crypto.randomBytes(32).toString('hex');
+        const { body: file } = await testServer.post(`/v2/buckets/${bucketId}/files/finish`)
+          .set('Authorization', getAuth(testUser))
+          .send({
+            index,
+            shards: [
+              { hash: crypto.randomBytes(20).toString('hex'), uuid: uploads[0].uuid, },
```
Comment (sg-gs) [inline tests/lib/e2e/buckets/files.e2e-spec.ts]: Same here, use array destructuring

File: tests/lib/e2e/users/users.e2e-spec.ts [hunk 16]
```diff
@@ -0,0 +1,360 @@
...
+            .send(payload)
+
+          // Assert
+          expect(response.status).toBe(201);
+          const users = await engine.storage.models.User.find({ _id: response.body.id })
+          expect(users).toHaveLength(1)
+
+          expect(users[0].toObject().activated).toBe(true)
+
+          // expect(dispatchSendGridMock).toHaveBeenCalled()
```
Comment (sg-gs) [inline tests/lib/e2e/users/users.e2e-spec.ts]: Remove comments

File: tests/lib/e2e/users/users.e2e-spec.ts [hunk 17]
```diff
@@ -0,0 +1,360 @@
...
+          // Arrange
+          const payload = { email: dataGenerator.email(), password: dataGenerator.hash({ length: 64 }), }
+          // Act: Create a user
+          const response = await testServer
+            .post('/v2/users')
+            .send(payload)
+
+          // Assert
+          expect(response.status).toBe(200);
+          // expect(response.status).toBe(201);
```
Comment (sg-gs) [inline tests/lib/e2e/users/users.e2e-spec.ts]: Remove comments

File: tests/lib/e2e/users/users.e2e-spec.ts [hunk 18]
```diff
@@ -0,0 +1,360 @@
...
+      describe('Deleting an existing user', () => {
+
+        it('When requesting user deactivation, it should work for authorized user', async () => {
+
+          // Arrange: Create User
+          const payload = { email: dataGenerator.email(), password: dataGenerator.hash({ length: 64 }), }
+          const { body: user } = await testServer
+            .post('/v2/users')
+            .send(payload)
+            // .expect(201)
```
Comment (sg-gs) [inline tests/lib/e2e/users/users.e2e-spec.ts]: Same here

File: tests/lib/e2e/utils.ts [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+  );
+
+  createdUser.password = user.password
+
+  return createdUser
+}
+
+export const deleteTestUser = async (args: Args = { }): Promise<void> => {
+  const { storage = engine.storage, user = testUser } = args
+  return await new Promise(resolve => storage.models.User.deleteOne({
```
Comment (sg-gs) [inline tests/lib/e2e/utils.ts]: return directly, using `async` and `await` here is not needed

File: tests/lib/e2e/utils.ts [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+    email: user.email,
+  }, (err: Error) => {
+    if (err) throw err
```
Comment (sg-gs) [inline tests/lib/e2e/utils.ts]: Do not throw, if you are using a Promise, instead, use the common reject flow

```
return new Promise((resolve, reject) => {
  // ....
  (err: Error) => {
    err ? reject(err) : resolve();
  }
})
```

**Review by sg-gs — CHANGES_REQUESTED** [review]

File: tests/lib/e2e/buckets/buckets.e2e-spec.ts (hunk 11 above)
Comment (sw-wayner) [reply tests/lib/e2e/buckets/buckets.e2e-spec.ts]: This is not necessary, global-teardown cleans up the test database after all tests. I'm removing it.

File: tests/lib/e2e/buckets/files.e2e-spec.ts [hunk 21]
```diff
@@ -0,0 +1,400 @@
...
+        const [firstUpload, secondUpload] = uploads;
+
+        expect(firstUpload.url).toBeNull();
+        expect(firstUpload.urls).toBeDefined();
+        expect(firstUpload.uuid).toBeDefined();
+        expect(firstUpload.UploadId).toBeDefined();
+
+        expect(secondUpload.url).toBeDefined();
+        expect(secondUpload.url).toBeNull();
+        expect(secondUpload.urls).toBeDefined();
```
Comment (sg-gs) [inline tests/lib/e2e/buckets/files.e2e-spec.ts]: This check may be not enough, as we should check if the `urls` length matches with the parts number + the URLs inside the array are in fact, URLs.

File: tests/lib/e2e/buckets/files.e2e-spec.ts [hunk 22]
```diff
@@ -0,0 +1,400 @@
...
+          .send({ uploads: fileParts, })
+
+        // Assert
+        expect(response.status).toBe(400)
+        expect(response.body.error).toBe('Multipart is not allowed for files smaller than 100MB')
+
+      })
+
+
+      it('When a user finishes to upload a file, the user can finish the upload with a hash per each part uploaded', async () => {
```
Comment (sg-gs) [inline tests/lib/e2e/buckets/files.e2e-spec.ts]: We are not testing the case of finishing a multipart upload

File: tests/lib/e2e/buckets/files.e2e-spec.ts [hunk 23]
```diff
@@ -0,0 +1,400 @@
...
+        const body = response.body;
+
+        expect(body.bucket).toBe(bucketId)
+        expect(body.created).toBeDefined()
+        expect(body.index).toBe(index)
+        expect(body.shards).toHaveLength(2)
+
+        const [firstShard, secondShard] = body.shards
+
+        expect(firstShard.hash).toBeDefined()
```
Comment (sg-gs) [inline tests/lib/e2e/buckets/files.e2e-spec.ts]: It could be defined, but null. We should check the value and the format of these data attributes

File: tests/lib/e2e/buckets/files.e2e-spec.ts [hunk 24]
```diff
@@ -0,0 +1,400 @@
...
+        // Act: remove the file
+        const response = await testServer.delete(`/buckets/${bucketId}/files/${file.id}`)
+          .set('Authorization', getAuth(testUser))
+
+        // Assert
+        expect(response.status).toBe(204);
+
+      })
+
+      it('When a user wants to delete a file, it should work if the file does not exist', async () => {
```
Comment (sg-gs) [inline tests/lib/e2e/buckets/files.e2e-spec.ts]: 'It should fail/It should return an error', isn't it?

File: tests/lib/e2e/buckets/files.e2e-spec.ts [hunk 25]
```diff
@@ -0,0 +1,400 @@
...
+        const { body } = response
+
+        expect(body.bucket).toBeDefined()
+        expect(body.operation).toBeDefined()
+        expect(body.expires).toBeDefined()
+        expect(body.token).toBeDefined()
+        expect(body.id).toBeDefined()
+        expect(body.encryptionKey).toBeDefined()
+        expect(body.mimetype).toBeDefined()
+        expect(body.size).toBeDefined()
```
Comment (sg-gs) [inline tests/lib/e2e/buckets/files.e2e-spec.ts]: Same here, let's validate the format of everything plus matching whatever is possible to be matched, in this case, the size, the bucket, the operation, etc.

File: tests/lib/e2e/buckets/files.e2e-spec.ts [hunk 26]
```diff
@@ -0,0 +1,400 @@
...
+      })
+
+      it('When a user wants to share a file, it should be able to get the file info', async () => {
```
Comment (sg-gs) [inline tests/lib/e2e/buckets/files.e2e-spec.ts]: This is not related to sharing directly. I would rename this legend to something related to being able to see the metadata of a file once is uploaded

**Review by sg-gs — CHANGES_REQUESTED** [review]

File: tests/lib/e2e/buckets/files.e2e-spec.ts (hunk 24 above)
Comment (wayn3r) [reply tests/lib/e2e/buckets/files.e2e-spec.ts]: Yes

**Review by sg-gs — APPROVED** [review]

**Comment by sg-gs** [comment]
Great work @sw-wayner
